### DISCUSSION TITLE
docs/v1.11.x/guide/vpc/

### DISCUSSION THREAD

**Discussion by giscus** [kubeovn/kube-ovn#2246] [Announcements]
# docs/v1.11.x/guide/vpc/

https://kubeovn.github.io/docs/v1.11.x/guide/vpc/

**Comment by ShaPoHun** [comment]
https://kubeovn.github.io/docs/v1.11.x/guide/vpc/#vpc_4 中，VpcNatGateway spec里面没有nextHopIP参数

**Reply by taohorse** [reply]
应该是指和静态路由中的nextHopIP参数一致

**Reply by oilbeater** [reply]
已添加

**Reply by tghfly** [reply]
VpcNatGateway  spec里也没有staticRoutes参数度

[root@master vm-test]# kubectl explain vpc-nat-gw  --recursive
KIND:     VpcNatGateway
VERSION:  kubeovn.io/v1

DESCRIPTION:
     <empty>

FIELDS:
   apiVersion   <string>
   kind <string>
   metadata     <Object>
      annotations       <map[string]string>
      creationTimestamp <string>
      deletionGracePeriodSeconds        <integer>
      deletionTimestamp <string>
      finalizers        <[]string>
      generateName      <string>
      generation        <integer>
      labels    <map[string]string>
      managedFields     <[]Object>
         apiVersion     <string>
         fieldsType     <string>
         fieldsV1       <map[string]>
         manager        <string>
         operation      <string>
         subresource    <string>
         time   <string>
      name      <string>
      namespace <string>
      ownerReferences   <[]Object>
         apiVersion     <string>
         blockOwnerDeletion     <boolean>
         controller     <boolean>
         kind   <string>
         name   <string>
         uid    <string>
      resourceVersion   <string>
      selfLink  <string>
      uid       <string>
   spec <Object>
      lanIp     <string>
      selector  <[]string>
      subnet    <string>
      tolerations       <[]Object>
         effect <string>
         key    <string>
         operator       <string>
         tolerationSeconds      <integer>
         value  <string>
      vpc       <string>

**Comment by lytos666** [comment]
在创建外部网络时，并没有配置与vpc-nat-gw路由的绑定选项，当有多个外部网络，是否支持特定的vpc-nat-gw绑定特定的外部网络。在多租户场景下，不同的租户给不同的vpc，不同的vpc创建各自的vpc-nat-gw，然后绑定各自的外部网络，这种方式是否支持？

**Reply by taohorse** [reply]
我觉得应该是可以支持的。

**Reply by oilbeater** [reply]
这个功能在我们的计划中，未来会进行支持

**Comment by Learntotolearn** [comment]
创建vpc网关后，自定义vpc中的pod还是无法访问互联网。
这是配置记录：
https://t.hitosea.com/single/file/OTIzOSwxOSxrd0ZNQmlabg==

**Reply by zbb88888** [reply]
https://kubeovn.github.io/docs/v1.12.x/guide/vpc/  Hyper-V及其他虚拟机下的网络（可能）无法直接使用macvlan连通Node网络。

**Comment by tghfly** [comment]
用的版本是v1.11.按文档中操作，test-vpc-1限制ns1使用，test-vpc-2限制ns2使用，但在创建pod时，都指定logical_switch为net1，都能分配IP，且能互ping通，请问，这个名称空间限制真有生效？
```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    ovn.kubernetes.io/logical_switch: net1
  namespace: ns1
  name: vpc1-pod
spec:
  containers:
    - name: vpc1-pod
      image: nginx:alpine
---
apiVersion: v1
kind: Pod
metadata:
  annotations:
    ovn.kubernetes.io/logical_switch: net1
  namespace: ns2
  name: vpc2-pod
spec:
  containers:
    - name: vpc2-pod
      image: nginx:alpine
```

[root@master vm-test]# kubectl get vpc
NAME          ENABLEEXTERNAL   STANDBY   SUBNETS                  NAMESPACES
ovn-cluster   false            true      ["join","ovn-default"]   
test-vpc-1    false            true      ["net1"]                 ["ns1"]
test-vpc-2    false            true      ["net2"]                 ["ns2"]
[root@master vm-test]# kubectl get subnet
NAME          PROVIDER   VPC           PROTOCOL   CIDR            PRIVATE   NAT     DEFAULT   GATEWAYTYPE   V4USED   V4AVAILABLE   V6USED   V6AVAILABLE   EXCLUDEIPS       U2OINTERCONNECTIONIP
join          ovn        ovn-cluster   IPv4       100.64.0.0/16   false     false   false     distributed   3        65530         0        0             ["100.64.0.1"]   
net1          ovn        test-vpc-1    IPv4       10.0.1.0/24     false     false   false     distributed   2        251           0        0             ["10.0.1.1"]     
net2          ovn        test-vpc-2    IPv4       10.0.1.0/24     false     false   false     distributed   0        253           0        0             ["10.0.1.1"]     
ovn-default   ovn        ovn-cluster   IPv4       12.244.0.0/16   false     true    true      distributed   58       65475         0        0             ["12.244.0.1"]  

[root@master vm-test]# kubectl get ips |grep net1
vpc1-pod.ns1                                                             10.0.1.5              00:00:00:A3:4B:15   node02   net1
vpc2-pod.ns2                                                             10.0.1.6              00:00:00:0F:41:F5   node02   net1
[root@master vm-test]# kubectl exec -it -n ns1 vpc1-pod -- ping 10.0.1.6
PING 10.0.1.6 (10.0.1.6): 56 data bytes
64 bytes from 10.0.1.6: seq=0 ttl=64 time=1.314 ms
64 bytes from 10.0.1.6: seq=1 ttl=64 time=0.837 ms
^C
--- 10.0.1.6 ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.837/1.075/1.314 ms

**Reply by zbb88888** [reply]
namespace: ns1 这种只是逻辑上的，你指定两个pod在同一子网，就是互通的。

**Reply by tghfly** [reply]
也就是说，只要知道subnet，即使不同的namespace仍然是可以设置相同的subnet来访问，这是设计就是如此吗？

**Reply by zbb88888** [reply]
如果是手动指定，目前确实如此。 如果需要严格要求pod的subnet必须属于pod所在的ns，应该还要加一些逻辑。

**Comment by qsliao** [comment]
containerd模式的K8S，镜像路径必须写全才行，比如kubeovn/vpc-nat-gateway:v1.11.3要写成docker.io/kubeovn/vpc-nat-gateway:v1.11.3才行

**Reply by oilbeater** [reply]
感谢支出，已更改

**Comment by kona3266** [comment]
```自定义 VPC 下的子网不支持默认 VPC 下的分布式网关和集中式网关```
这句话中的默认vpc是指ovn-cluster这个vpc吗，分布式网关和集中式网关是指什么呢？

**Reply by kona3266** [reply]
找到了，原句应该是指不支持overlay模式下的子网设置gatewayType字段。
